File: server/services/stats.js
import { getDb } from '../db/index.js'

const buffer = new Map() // key: JSON.stringify(row) -> row sums
let active = 0
let lastFlush = Date.now()

// ponytail: bucket on model_id (FK) when present; fall back to (provider_id, model_name) for legacy rows.
// Single global lock (JS event loop) — fine for single-node deployment.
export function acquire() { active++ }
export function release() { if (active > 0) active-- }
export function getActive() { return active }

function bucketKey(r) {
  return `${r.api_key_id ?? ''}|${r.model_id ?? ''}|${r.provider_id ?? ''}|${r.model_name}`
}

export function record(r) {
  const k = bucketKey(r)
  const cur = buffer.get(k)
  if (cur) {
    cur.request_count += r.request_count ?? 1
    cur.input_tokens += r.input_tokens ?? 0
    cur.output_tokens += r.output_tokens ?? 0
    cur.cached_tokens += r.cached_tokens ?? 0
  } else {
    buffer.set(k, {
      api_key_id: r.api_key_id ?? null,
      provider_id: r.provider_id ?? null,
      model_id: r.model_id ?? null,
      model_name: r.model_name,
      request_count: r.request_count ?? 1,
      input_tokens: r.input_tokens ?? 0,
      output_tokens: r.output_tokens ?? 0,
      cached_tokens: r.cached_tokens ?? 0,
    })
  }
  if (Date.now() - lastFlush > 10000) flush()
}

export async function flush() {
  if (buffer.size === 0) return
  const rows = [...buffer.values()]
  buffer.clear()
  lastFlush = Date.now()
  const db = getDb()
  for (const r of rows) {
    await db.run(
      'INSERT INTO usage_stats(api_key_id, provider_id, model_id, model_name, request_count, input_tokens, output_tokens, cached_tokens) VALUES (?,?,?,?,?,?,?,?)',
      [r.api_key_id, r.provider_id, r.model_id ?? null, r.model_name, r.request_count, r.input_tokens, r.output_tokens, r.cached_tokens],
    )
  }
}

let interval = null
export function startFlushTask() {
  if (interval) return
  interval = setInterval(() => { flush().catch(() => {}) }, 10000)
}
export function stopFlushTask() {
  if (interval) clearInterval(interval)
  interval = null
}

export async function query({ provider_id, model_name, api_key_id, range }) {
  // ponytail: time-range filter so Dashboard/Stats show 'today' vs '7d' etc.
  // Use localtime to respect user timezone (SQLite date('now') is UTC).
  const where = []
  const params = []
  if (provider_id) { where.push('u.provider_id=?'); params.push(provider_id) }
  if (model_name) { where.push('u.model_name=?'); params.push(model_name) }
  if (api_key_id) { where.push('u.api_key_id=?'); params.push(api_key_id) }
  if (range === 'today') { where.push("datetime(u.created_at,'localtime') >= date('now','localtime')"); }
  else if (range === '7d') { where.push("datetime(u.created_at,'localtime') >= date('now','-7 days','localtime')"); }
  else if (range === '30d') { where.push("datetime(u.created_at,'localtime') >= date('now','-30 days','localtime')"); }
  const w = where.length ? `WHERE ${where.join(' AND ')}` : ''

  // ponytail: breakdown keeps per-key detail (admin table);
    // topModels/topProviders are pre-aggregated to dedupe same provider+model across multiple keys (#4).
    // v0.2.14: LEFT JOIN via model_id (FK) — picks up renamed models automatically.
    // For legacy rows (model_id IS NULL), fall back to (provider_id, model_name) lookup to keep price stats.
    const rows = await getDb().query(
      `SELECT p.name AS provider, u.model_name, u.api_key_id, k.name AS key_name,
              COALESCE(m.model_name, u.model_name) AS current_model_name,
              COALESCE(m.input_price, m_legacy.input_price, 0) AS input_price,
              COALESCE(m.output_price, m_legacy.output_price, 0) AS output_price,
              COALESCE(m.cache_hit_price, m_legacy.cache_hit_price, 0) AS cache_hit_price,
              SUM(u.request_count) AS request_count,
              SUM(u.input_tokens) AS input_tokens,
              SUM(u.output_tokens) AS output_tokens,
              SUM(u.cached_tokens) AS cached_tokens
       FROM usage_stats u
       LEFT JOIN providers p ON u.provider_id = p.id
       LEFT JOIN api_keys k ON u.api_key_id = k.id
       LEFT JOIN models m ON u.model_id = m.id
       LEFT JOIN models m_legacy ON u.model_id IS NULL AND u.provider_id = m_legacy.provider_id AND u.model_name = m_legacy.model_name
       ${w}
       GROUP BY u.provider_id, u.model_name, u.api_key_id
       ORDER BY request_count DESC`,
      params,
    )

  // ponytail: pre-aggregate by (provider, model) so the Top-5 panel doesn't show the same
  // provider/model multiple times when several API keys used it. One pass over `rows`.
  // Cost uses (input-cached)*input_price + output*output_price + cached*cache_hit_price (v0.2.7 Bug 24).
  const rowCost = (r) =>
    ((r.input_tokens || 0) - (r.cached_tokens || 0)) * (r.input_price || 0) / 1_000_000
    + (r.output_tokens || 0) * (r.output_price || 0) / 1_000_000
    + (r.cached_tokens || 0) * (r.cache_hit_price || 0) / 1_000_000
  const byModel = new Map()
  const byProvider = new Map()
  for (const r of rows) {
    const mk = `${r.provider}|${r.current_model_name || r.model_name}`
    const m = byModel.get(mk) || { provider: r.provider || '?', model: r.current_model_name || r.model_name || '?',
      request_count: 0, input_tokens: 0, output_tokens: 0, cached_tokens: 0,
      input_price: r.input_price || 0, output_price: r.output_price || 0, cache_hit_price: r.cache_hit_price || 0,
      cost: 0 }
    m.request_count += r.request_count || 0
    m.input_tokens += r.input_tokens || 0
    m.output_tokens += r.output_tokens || 0
    m.cached_tokens += r.cached_tokens || 0
    m.cost += rowCost(r)
    byModel.set(mk, m)

    const pk = r.provider || '?'
    // ponytail: don't carry a single model's price into the provider aggregate — different models
    // within the same provider have different prices (e.g. Deepseek v4-pro vs v4-flash). Sum cost
    // row-by-row using each row's own price; expose cost on the aggregate, drop the misleading price.
    const p = byProvider.get(pk) || { provider: pk, request_count: 0,
      input_tokens: 0, output_tokens: 0, cached_tokens: 0, cost: 0 }
    p.request_count += r.request_count || 0
    p.input_tokens += r.input_tokens || 0
    p.output_tokens += r.output_tokens || 0
    p.cached_tokens += r.cached_tokens || 0
    p.cost += rowCost(r)
    byProvider.set(pk, p)
  }
  const topModels = [...byModel.values()].sort((a, b) => (b.input_tokens + b.output_tokens) - (a.input_tokens + a.output_tokens)).slice(0, 5)
  const topProviders = [...byProvider.values()].sort((a, b) => (b.input_tokens + b.output_tokens) - (a.input_tokens + a.output_tokens)).slice(0, 5)

  let total_requests = 0, total_input_tokens = 0, total_output_tokens = 0, total_cached_tokens = 0
  for (const r of rows) {
    total_requests += r.request_count || 0
    total_input_tokens += r.input_tokens || 0
    total_output_tokens += r.output_tokens || 0
    total_cached_tokens += r.cached_tokens || 0
  }
  return { total_requests, total_input_tokens, total_output_tokens, total_cached_tokens, breakdown: rows, topModels, topProviders }
}

File: server/internal/stats/stats.go
package stats

import (
	"fmt"
	"log"
	"sort"
	"strings"
	"sync"
	"time"

	"server-go/internal/db"
)

type RecordInput struct {
	APIKeyID     int64
	ProviderID   int64
	ModelID      int64
	ModelName    string
	RequestCount int
	InputTokens  int
	OutputTokens int
	CachedTokens int
	Cost         float64 // ponytail: v0.4.24 — USD at request time, written once, immune to later price edits
}

type bucketKey string

var (
	buffer    = sync.Map{}
	lastFlush = time.Now()
	mu        sync.Mutex
)

func Acquire() { /* not used in Go impl */ }
func Release() { /* not used in Go impl */ }

func key(r RecordInput) bucketKey {
	return bucketKey(fmt.Sprintf("%d|%d|%d|%s", r.APIKeyID, r.ModelID, r.ProviderID, r.ModelName))
}

func Record(r RecordInput) {
	k := key(r)
	mu.Lock()
	if v, ok := buffer.Load(k); ok {
		cur := v.(*RecordInput)
		cur.RequestCount += max(r.RequestCount, 1)
		cur.InputTokens += r.InputTokens
		cur.OutputTokens += r.OutputTokens
		cur.CachedTokens += r.CachedTokens
		cur.Cost += r.Cost
	} else {
		cp := r
		if cp.RequestCount == 0 {
			cp.RequestCount = 1
		}
		buffer.Store(k, &cp)
	}
	if time.Since(lastFlush) > 10*time.Second {
		go Flush()
	}
	mu.Unlock()
}

func Flush() {
	mu.Lock()
	var rows []*RecordInput
	buffer.Range(func(k, v interface{}) bool {
		rows = append(rows, v.(*RecordInput))
		buffer.Delete(k)
		return true
	})
	lastFlush = time.Now()
	mu.Unlock()
	if len(rows) == 0 {
		return
	}
	d := db.Get()
	for _, r := range rows {
		if _, err := d.Run("INSERT INTO usage_stats(api_key_id, provider_id, model_id, model_name, request_count, input_tokens, output_tokens, cached_tokens, cost) VALUES (?,?,?,?,?,?,?,?,?)",
			r.APIKeyID, r.ProviderID, r.ModelID, r.ModelName, r.RequestCount, r.InputTokens, r.OutputTokens, r.CachedTokens, r.Cost); err != nil {
			log.Printf("[stats] flush insert error: %v", err)
		}
	}
}

var flushTicker *time.Ticker

func StartFlushTask() {
	if flushTicker != nil {
		return
	}
	flushTicker = time.NewTicker(10 * time.Second)
	go func() {
		for range flushTicker.C {
			Flush()
		}
	}()
}

func StopFlushTask() {
	if flushTicker != nil {
		flushTicker.Stop()
	}
}

type StatsQuery struct {
	ProviderID int64
	ModelName  string
	APIKeyID   int64
	Range      string
	Capability string // ponytail: empty = all; "chat" or "embedding"
}

func toFloat(v interface{}) float64 {
	switch n := v.(type) {
	case float64:
		return n
	case int64:
		return float64(n)
	case int:
		return float64(n)
	}
	return 0
}

func Query(q StatsQuery) (map[string]interface{}, error) {
	d := db.Get()
	where := []string{}
	params := []interface{}{}
	if q.ProviderID > 0 {
		where = append(where, "u.provider_id=?")
		params = append(params, q.ProviderID)
	}
	if q.ModelName != "" {
		where = append(where, "u.model_name=?")
		params = append(params, q.ModelName)
	}
	if q.APIKeyID > 0 {
		where = append(where, "u.api_key_id=?")
		params = append(params, q.APIKeyID)
	}
	if q.Capability != "" {
		// ponytail: COALESCE(m.capability, m_legacy.capability, 'chat') AS capability
		where = append(where, "COALESCE(m.capability, m_legacy.capability, 'chat') = ?")
		params = append(params, q.Capability)
	}
	switch q.Range {
	case "today":
		where = append(where, "datetime(u.created_at,'localtime') >= date('now','localtime')")
	case "7d":
		where = append(where, "datetime(u.created_at,'localtime') >= date('now','-7 days','localtime')")
	case "30d":
		where = append(where, "datetime(u.created_at,'localtime') >= date('now','-30 days','localtime')")
	}
	w := ""
	if len(where) > 0 {
		w = "WHERE " + strings.Join(where, " AND ")
	}

	sql := "SELECT p.name AS provider, u.model_name, u.api_key_id, k.name AS key_name, " +
		"COALESCE(m.model_name, u.model_name) AS current_model_name, " +
		"COALESCE(m.input_price, m_legacy.input_price, 0) AS input_price, " +
		"COALESCE(m.output_price, m_legacy.output_price, 0) AS output_price, " +
		"COALESCE(m.cache_hit_price, m_legacy.cache_hit_price, 0) AS cache_hit_price, " +
		"COALESCE(m.capability, m_legacy.capability, 'chat') AS capability, " +
		"SUM(u.request_count) AS request_count, " +
		"SUM(u.input_tokens) AS input_tokens, " +
		"SUM(u.output_tokens) AS output_tokens, " +
		"SUM(u.cached_tokens) AS cached_tokens, " +
		"SUM(u.cost) AS cost " +
		"FROM usage_stats u " +
		"LEFT JOIN providers p ON u.provider_id = p.id " +
		"LEFT JOIN api_keys k ON u.api_key_id = k.id " +
		"LEFT JOIN models m ON u.model_id = m.id " +
		"LEFT JOIN models m_legacy ON u.model_id IS NULL AND u.provider_id = m_legacy.provider_id AND u.model_name = m_legacy.model_name " +
		w + " " +
		"GROUP BY u.provider_id, u.model_name, u.api_key_id " +
		"ORDER BY request_count DESC"

	rows, err := d.Query(sql, params...)
	if err != nil {
		return nil, err
	}

	// ponytail: v0.4.24 — cost is summed from stored usage_stats.cost (price at request time),
	// not recomputed from current model prices. Recomputing would silently rewrite history
	// every time a provider edits a price.

	type aggData struct {
		Req, In, Out, Cached                           int64
		Cost                                             float64
		InputPrice, OutputPrice, CachePrice              float64
	}
	byModel := map[string]*aggData{}
	byProvider := map[string]*aggData{}
	for _, r := range rows {
		mk := db.Str(r["provider"]) + "|" + db.Str(r["current_model_name"])
		m := byModel[mk]
		if m == nil {
			m = &aggData{InputPrice: toFloat(r["input_price"]), OutputPrice: toFloat(r["output_price"]), CachePrice: toFloat(r["cache_hit_price"])}
			byModel[mk] = m
		}
		m.Req += db.Int64(r["request_count"])
		m.In += db.Int64(r["input_tokens"])
		m.Out += db.Int64(r["output_tokens"])
		m.Cached += db.Int64(r["cached_tokens"])
		m.Cost += toFloat(r["cost"])

		pk := db.Str(r["provider"])
		if pk == "" {
			pk = "?"
		}
		pa := byProvider[pk]
		if pa == nil {
			pa = &aggData{}
			byProvider[pk] = pa
		}
		pa.Req += db.Int64(r["request_count"])
		pa.In += db.Int64(r["input_tokens"])
		pa.Out += db.Int64(r["output_tokens"])
		pa.Cached += db.Int64(r["cached_tokens"])
		pa.Cost += toFloat(r["cost"])
	}

	var totalReq, totalIn, totalOut, totalCached int64
	var totalCost float64
	for _, r := range rows {
		totalReq += db.Int64(r["request_count"])
		totalIn += db.Int64(r["input_tokens"])
		totalOut += db.Int64(r["output_tokens"])
		totalCached += db.Int64(r["cached_tokens"])
		totalCost += toFloat(r["cost"])
	}

	topModels := make([]map[string]interface{}, 0)
	for k, v := range byModel {
		parts := strings.SplitN(k, "|", 2)
		prov, model := parts[0], ""
		if len(parts) > 1 {
			model = parts[1]
		}
		topModels = append(topModels, map[string]interface{}{
			"provider": prov, "model": model,
			"request_count": v.Req, "input_tokens": v.In,
			"output_tokens": v.Out, "cached_tokens": v.Cached,
			"cost": v.Cost, "input_price": v.InputPrice,
			"output_price": v.OutputPrice, "cache_hit_price": v.CachePrice,
		})
	}
	sort.Slice(topModels, func(i, j int) bool {
		ri := db.Int64(topModels[i]["request_count"])
		rj := db.Int64(topModels[j]["request_count"])
		return ri > rj
	})
	if len(topModels) > 5 {
		topModels = topModels[:5]
	}

	topProviders := make([]map[string]interface{}, 0)
	for k, v := range byProvider {
		topProviders = append(topProviders, map[string]interface{}{
			"provider": k, "request_count": v.Req,
			"input_tokens": v.In, "output_tokens": v.Out,
			"cached_tokens": v.Cached, "cost": v.Cost,
		})
	}
	sort.Slice(topProviders, func(i, j int) bool {
		ri := db.Int64(topProviders[i]["request_count"])
		rj := db.Int64(topProviders[j]["request_count"])
		return ri > rj
	})
	if len(topProviders) > 5 {
		topProviders = topProviders[:5]
	}

	return map[string]interface{}{
		"total_requests":    totalReq,
		"total_input_tokens":  totalIn,
		"total_output_tokens": totalOut,
		"total_cached_tokens": totalCached,
		"total_cost":         totalCost,
		"breakdown":    rows,
		"topModels":    topModels,
		"topProviders": topProviders,
	}, nil
}
